Add a Google Cloud PubSub-based implementation of PCollectionCache

ostrokach opened the pull request:
Add a `PubSubBasedCache` implementation of the `PCollectionCache` ABC. 

Demos:

| Notebook | MyBinder | Colab |
| ------------- | ------------- | -------- |
| [03-streaming_demo_explicit_caching.ipynb](https://github.com/ostrokach/beam-notebooks/blob/master/notebooks/03-streaming_demo_explicit_caching.ipynb) | [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/ostrokach/beam-notebooks/master?filepath=notebooks%2F03-streaming_demo_explicit_caching.ipynb) | [![Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ostrokach/beam-notebooks/blob/master/notebooks/03-streaming_demo_explicit_caching.ipynb)
| [04-streaming_demo_explicit_caching_taxirides.ipynb](https://github.com/ostrokach/beam-notebooks/blob/master/notebooks/04-streaming_demo_explicit_caching_taxirides.ipynb) | [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/ostrokach/beam-notebooks/master?filepath=notebooks%2F04-streaming_demo_explicit_caching_taxirides.ipynb) | [![Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ostrokach/beam-notebooks/blob/master/notebooks/04-streaming_demo_explicit_caching_taxirides.ipynb)

CC: @ivant @ananvay  
R: @aaltay @udim 

-----------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

ostrokach commented:
retest this please

ostrokach commented:
retest this please

ostrokach commented:
retest this please

ostrokach commented:
retest this please

aaltay commented:
R: @KevinGG could you help with reviewing this PR?

udim commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+        return self.timestamp < other.timestamp
+
+    # Set arbitrary queue size limit to prevent OOM errors.
+    parsed_message_queue = queue.PriorityQueue(1000)
```
Why a PriorityQueue? I don't think it can guarantee that messages yielded will have non-decreasing timestamp values.

ostrokach replied:
From my understanding, it does guarantee this. I had to implement the `PrioritizedTimestampedValue` class above so that values are yielded in non-decreasing order based on their timestamps rather than their values.

udim commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+    # Initialize PubSub resources. Note that we cannot save pub_client and
+    # sub_client as class attributes because they fail to serialize using
+    # pickle.
+    pub_client = pubsub.PublisherClient()
```
Ideally the code that reads and writes to/from Pub/Sub would be reused. Specifically I'm thinking about the DirectRunner code, like the `rfc3339_or_milli` code.

ostrokach replied:
I agree. However, I think it may be better to leave this as a TODO to be addressed in a future refactoring? Since InteractiveRunner is alpha, the code for caching to PubSub is likely to undergo several iterations. If those changes also affect DirectRunner, it may increase the surface area for introducing bugs or at least slow the development cycle?

udim commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+
+  def process(self, message):
+    # pylint: disable=reimported
+    from apache_beam.transforms.window import TimestampedValue
```
Could you explain why the process methods have imports in them?

ostrokach replied:
Sorry, I have gotten used to doing this when writing DoFns inside scripts / notebooks with `save_main_session=False`.

udim commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+  pass
+
+
+class PubSubBasedCache(StreamBasedCache):
```
Could you an explanation on how this cache works?
- How cached data is encoded.
- Where cached data is written to.
- How snapshots are used. What happens when snapshots are supported, and what happens when they are not.
- What with_attributes does.
- The special "ts" attribute.

ostrokach replied:
I covered each of those points in the docstring for `PubSubBasedCache`. Thanks for the detailed list!

udim commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+  }
+  requires_coder = True
+
+  _default_timestamp_attribute = "ts"
```
I believe our style is to have constants in all-caps. I might be wrong though. /shrug

ostrokach replied:
"ts" is the name of the timestamp attribute in the `projects/pubsub-public-data/topics/taxirides-realtime` PubSub topic that is used in the [NYC Taxi Tycoon](https://codelabs.developers.google.com/codelabs/cloud-dataflow-nyc-taxi-tycoon/#3). However, I not have much experience with real-world conventions.

udim commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+    writer_kwargs = self._writer_kwargs.copy()
+    attributes_fn = writer_kwargs.pop("attributes_fn", None)
+    # DirectRunner does not support timestamp_attribute
+    timestamp_attribute = writer_kwargs.pop("timestamp_attribute", "ts")
```
Please use the _default_timestamp_attribute constant.

ostrokach replied:
Done.

udim commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+      self.coder = coders.registry.get_coder(element_type)
+
+    writer_kwargs = self._writer_kwargs.copy()
+    attributes_fn = writer_kwargs.pop("attributes_fn", None)
```
What is this keyword argument? I couldn't find documentation for it.

ostrokach replied:
I added a docstring which includes the description of this argument. Please let me know if you think it is not necessary and should be removed.

udim commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+    # DirectRunner does not support timestamp_attribute
+    timestamp_attribute = writer_kwargs.pop("timestamp_attribute")
+
+    if writer_kwargs.get("with_attributes"):
```
I think this code could be a lot simpler Pub/Sub message data was the element and the attributes contained any additional metadata used by the cache. I don't understand what `attributes_fn` is used for (and where is it set?).

ostrokach replied:
In order to be consistent with `WriteToPubSub` and `ReadFromPubSub`, *with_attributes* has to be a boolean flag indicating whether the input elements are bytes (*with_attributes=False*) or instances of `PubsubMessage` (*with_attributes = True*).

Forcing the user to manually convert their data into PubsubMessages is a bad user experience because, in the case of caching, the user probably just wants to examine the elements in the stream without applying any additional PTransforms. Therefore, I think *with_attributes* should be False in most cases. However, it is possible that the user wants to write their data to cache with a specific timestamp inferred from the data, rather than the event time of a given element. In order to make this possible for any data type, I added an `attributes_fn` argument, which takes a function that maps a given element to a timestamp.

If you think it would be simpler to get rid of this additional argument, please confirm and I will do so.

ostrokach replied:
I cleaned up the code around `PatchedPubSubReader` and `PatchedPubSubWriter` so hopefully it should be a bit easier to read.

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+      assert created_subsciption in self._child_subscriptions
+
+    @functools.total_ordering
+    class PrioritizedTimestampedValue(TimestampedValue):
```
`TimestampedValue` does implement comparisons, but those comparisons are done primarily based on the value rather than the timestamp. Here, we want the values with the oldest timestamp to come first. Would you think that changing the name of this class would make the intention more clear?

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/filebasedcache_it_test.py:
```diff
@@ -0,0 +1,317 @@
+
+  test_data = [{"a": 11, "b": "XXX"}, {"a": 20, "b": "YYY"}]
+
+  def check_serde_empty(self, write_fn, read_fn, serializer):
```
I renamed those functions to `{check,test}_serialize_deserialize_{empty,filled}`.

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/filebasedcache_it_test.py:
```diff
@@ -0,0 +1,317 @@
+def validate_through_pipeline(cache, expected):
+
+  def equal_to_expected(actual):
+    test = TestCase("__init__")
```
OK, I will do that, thanks! I could not think of a better way.

ostrokach replied:
On Python 2.7, I had to add the following code in order to get this to work. It is still ugly, but I guess less ugly than what I had before:

```python
    ...

    if sys.version_info < (3,):

      def runTest(self):
        pass
```

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+  def expand(self, pbegin):
+    reader_kwargs = self._reader_kwargs.copy()
+
+    reader = self._reader_class(*self._reader_args, **reader_kwargs)
```
I replaced _reader_class with ReadFromPubSub in PatchedPubSubReader and _writer_class with WriteToPubSub in PatchedPubSubWriter.

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/pcollectioncache.py:
```diff
@@ -0,0 +1,135 @@
+
+  @property
+  @abc.abstractmethod
+  def _reader_class(self):
```
Done!

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/filebasedcache.py:
```diff
@@ -0,0 +1,276 @@
+#
```
Done!

ostrokach commented on sdks/python/setup.py:
```diff
@@ -104,6 +104,7 @@ def get_version():
 REQUIRED_PACKAGES = [
     'avro>=1.8.1,<2.0.0; python_version < "3.0"',
     'avro-python3>=1.8.1,<2.0.0; python_version >= "3.0"',
+    'backports.weakref>=1.0,<1.1',
```
Done!

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache_it_test.py:
```diff
@@ -0,0 +1,214 @@
+def validate_directly(cache, expected):
+  """Validate the contents of a cache by reading it directly."""
+  output = list(itertools.islice(cache.read(timeout=10), len(expected)))
+  actual = [o.value for o in output]
```
Yes, thank you!

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/pcollectioncache.py:
```diff
@@ -0,0 +1,135 @@
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def write(self):
```
Done!

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/pcollectioncache.py:
```diff
@@ -0,0 +1,135 @@
+
+  @property
+  @abc.abstractmethod
+  def persist(self):
```
Persistence is useful in cases where we want to restart the Python interpreter / notebook in which we are working (for example, we ran out of memory, want to re-import some package, etc.). It may be nice in some cases to be able to reload the caches inside the new session.

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py:
```diff
@@ -0,0 +1,491 @@
+      raise ValueError(
+          "timestamp_attribute must be specified when reading from cache.")
+
+  def expand(self, pbegin):
```
Done!

ostrokach commented on sdks/python/apache_beam/runners/interactive/caching/streambasedcache_it_test.py:
```diff
@@ -0,0 +1,214 @@
+    pub_client.delete_topic(introspect_topic.name)
+
+
+def retry_flakes(fn):
```
I don't have much experience with this. My rationale was that, since InteractiveRunner is in alpha, we want to avoid breaking builds with flaky InteractiveRunner tests. And since PubSub can deliver the same message more than once, those tests are guaranteed to fail sooner or later.

ostrokach commented:
Hi @udim, thanks a lot for the detailed reviews or the two PRs! I think I have addressed most of your concerns on this PR as well. However, I forgot to make a copy of existing changes before force-pushing, so I can't seem to make a diff :( (sorry!).

Please let me know if you see other things that need to be changed.

Thanks!

aaltay commented:
There is a quite a bit of an overlap between this and https://github.com/apache/beam/pull/8884 . Perhaps the other PR should be reviewed first.

/cc @KevinGG

ostrokach commented:
@stale I'm still working on this.

ostrokach commented:
retest this please
